api/server.go

```diff
@@ -0,0 +1,299 @@
+	Status      int
+	Data        interface{}
+	RedirectURL string
+	Error       error
+}
+
+// Handler handles API requests
+type Handler func(r *http.Request) Response
+
+func (h Handler) callHandlerSafely(r *http.Request) (rsp Response) {
+	defer func() {
+		if r := recover(); r != nil {
+			err, ok := r.(error)
+			if !ok {
+				err = errors.Err("%v", r)
+			}
+			rsp = Response{Error: errors.Wrap(err, 2)}
+		}
+	}()
+
+	return h(r)
+}
+
+func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Set header settings
+	if HeaderSettings != nil {
+		//Multiple readers, no writers is okay
+		for key, value := range *HeaderSettings {
+			w.Header().Set(key, value)
+		}
+	}
+
+	// Stop here if its a preflighted OPTIONS request
+	if r.Method == "OPTIONS" {
+		return
+	}
+
+	rsp := h.callHandlerSafely(r)
+
+	if rsp.Status == 0 {
+		if rsp.Error != nil {
+			if statusError, ok := rsp.Error.(StatusError); ok {
+				rsp.Status = statusError.Status
+			} else if errors.Is(rsp.Error, errAuthenticationRequired) {
+				rsp.Status = http.StatusUnauthorized
+			} else if errors.Is(rsp.Error, errNotAuthenticated) || errors.Is(rsp.Error, errForbidden) {
+				rsp.Status = http.StatusForbidden
+			} else {
+				rsp.Status = http.StatusInternalServerError
+			}
+		} else if rsp.RedirectURL != "" {
+			rsp.Status = http.StatusFound
+		} else {
+			rsp.Status = http.StatusOK
+		}
+	}
+
+	success := rsp.Status < http.StatusBadRequest
+
+	consoleText := r.RemoteAddr + " [" + strconv.Itoa(rsp.Status) + "]: " + r.Method + " " + r.URL.Path
+	// query := r.URL.Query()
+	// if len(query) > 0 {
+	// 	queryJSON, _ := json.Marshal(query)
+	// 	consoleText += " " + string(queryJSON)
+	// }
+	if success {
+		log.Println(color.GreenString(consoleText))
+	} else {
+		log.Errorln(color.RedString(consoleText))
+	}
+
+	// redirect
+	if rsp.Status >= http.StatusMultipleChoices && rsp.Status < http.StatusBadRequest {
+		http.Redirect(w, r, rsp.RedirectURL, rsp.Status)
+		return
+	} else if rsp.RedirectURL != "" {
+		// warn us that we messed up somehow
+	}
+
+	var errorString *string
+	if rsp.Error != nil {
+		errorStringRaw := rsp.Error.Error()
+		errorString = &errorStringRaw
+	}
+
+	var trace []string
+	if TraceEnabled && errors.HasTrace(rsp.Error) {
+		trace = strings.Split(errors.Trace(rsp.Error), "\n")
+		for index, element := range trace {
+			if strings.HasPrefix(element, "\t") {
+				trace[index] = strings.Replace(element, "\t", "    ", 1)
+			}
+		}
+	}
+
+	// http://choly.ca/post/go-json-marshalling/
+	jsonResponse, err := json.MarshalIndent(&struct {
+		Success bool        `json:"success"`
+		Error   *string     `json:"error"`
+		Data    interface{} `json:"data"`
+		Trace   []string    `json:"_trace,omitempty"`
+	}{
+		Success: success,
+		Error:   errorString,
+		Data:    rsp.Data,
+		Trace:   trace,
+	}, "", "  ")
+	if err != nil {
+		errMsg := "Error encoding JSON response: " + err.Error()
+		log.Errorln(errMsg)
+		ErrorHandler(errors.Base(errMsg))
```

Comment:
when you do this, you're losing the trace of the original error. you should probably change this whole section to

```go
if err != nil {
  ErrorHandler(errors.Prefix("error encoding JSON response", err))
}
```